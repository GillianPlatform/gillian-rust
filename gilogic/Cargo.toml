[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }