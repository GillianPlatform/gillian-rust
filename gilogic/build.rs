fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
