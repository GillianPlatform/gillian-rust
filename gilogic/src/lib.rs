#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
