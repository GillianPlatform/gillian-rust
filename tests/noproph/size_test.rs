//@check-pass
extern crate gilogic;

use gilogic::{macros::*, Ownable};

struct Test {
    x: u32,
    y: u16,
}

#[show_safety]
fn test() {
    assert!(std::mem::size_of::<Test>() >= 6);
}
