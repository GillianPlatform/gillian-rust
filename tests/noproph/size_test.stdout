#begin_init_data
{"Test":["Struct","ReprRust",[["x",["Scalar","u32"]],["y",["Scalar","u16"]]]]}
#end_init_data

import "i__binop.gil", "i__lang.gil", "i__std_shims.gil", "i__prophecies.gil", "i__std_shims_no_pcy.gil", "i__lang_no_pcy.gil";

pred "<() as gilogic::Ownable>::own"(+self):
  (#self == {{  }}) * (self == #self);


spec test()
[[ emp ]]
[[ "<() as gilogic::Ownable>::own"(ret) ]]
normal

proc test() {
  bb0:  g_0 := "std::mem::size_of"({{ "adt", "Test", {{  }} }});
        m_3 := g_0;
        goto bb1;
  bb1:  g_1 := m_3;
        m_2 := ((6i) i<= (g_1));
        g_2 := m_2;
        goto [g_2] bb2 bb3;
  bb2:  m_1 := {{}};
        m_0 := {{}};
        g_3 := m_0;
        ret := g_3;
        goto rlab;
  bb3:  g_4 := [new_lft]();
        pLft_2 := l-nth(g_4, 0i);
        g_5 := "core::panicking::panic"(pLft_0, "assertion failed: std::mem::size_of::<Test>() >= 6");
        m_4 := g_5;
  rlab: return
};

